### PR TITLE
[dev] mHC kernel fusion 

### PR DESCRIPTION
**Pull request by jingqiny-99**
# What does this PR do ?

kernel fusion (using cuTile) for mHC.

Verified using 27B Model (introduced in original mHC paper) with GB200

Throughput: 
image

Validation LM Loss:
image

Iteration Time:
image

### Config: `s=8192, b=4, n=4, C=7168`

| Kernel | Framework | Fwd (ms) | Bwd (ms) | E2E (ms) | Speedup |
|--------|-----------|---------:|---------:|---------:|--------:|
| sinkhorn | compiled | 0.346 | 1.018 | 1.364 | 1.00x |
| sinkhorn | cutile | 0.065 | 0.133 | 0.198 | 6.89x |
| h_aggregate | compiled | 0.338 | 0.808 | 1.146 | 1.00x |
| h_aggregate | cutile | 0.349 | 0.661 | 1.010 | 1.13x |
| h_post_bda | compiled | 3.829 | 6.663 | 10.492 | 1.00x |
| h_post_bda | cutile | 0.684 | 2.549 | 3.234 | 3.24x |
| proj_rms | compiled | 0.554 | 2.204 | 2.758 | 1.00x |
| proj_rms | cutile | 0.426 | 1.541 | 1.967 | 1.40x |

## Prerequisites

Linux x86_64, Linux aarch64 or Windows x86_64

A GPU with compute capability 8.x 10.x, 11.x or 12.x

NVIDIA Driver r580 or later

Python version 3.10, 3.11, 3.12 or 3.13

## Usage

Install CuTile: `pip install cuda-tile[tileiras]`

Enable fusion kernel with `--use-fused-mhc`

### Pre-checks

- [x] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

File: megatron/core/fusions/fused_mhc_kernels.py [hunk 1]
```diff
@@ -0,0 +1,1090 @@
...
+            g_x_2d = ct.full((1, TILE_C), 0, dtype=hp.dtype)
+            g_orig_2d = ct.full((N, TILE_C), 0, dtype=hp.dtype)
+            for j in range(N):
+                g_x_2d += ct.extract(hp_2d, (j, 0), shape=(1, 1)).item() * ct.extract(
+                    go_2d, (j, 0), shape=(1, TILE_C)
+                )
+                g_orig_2d += ct.extract(hr_2d, (j, 0), shape=(1, N)).reshape((N, 1)) * ct.extract(
+                    go_2d, (j, 0), shape=(1, TILE_C)
+                )
+            acc_g_hp_2d += ct.sum(go_2d * xb_2d, axis=1, keepdims=True)
```
[the remark this answers is not in the document]
Comment (jingqiny-99) [reply megatron/core/fusions/fused_mhc_kernels.py:415]: Fixed

File: megatron/core/fusions/fused_mhc_kernels.py [hunk 2]
```diff
@@ -0,0 +1,1090 @@
...
+
+# -- Sinkhorn ----------------------------------------------------------------
+def _ref_sinkhorn_fwd(input_logits: Tensor, num_iterations: int, eps: float = 1e-8) -> Tensor:
+    M = torch.exp(input_logits)
+    for _ in range(num_iterations):
+        M = M / M.sum(dim=-1, keepdim=True).clamp(min=eps)
+        M = M / M.sum(dim=-2, keepdim=True).clamp(min=eps)
+    return M
+
+
```
[the remark this answers is not in the document]
Comment (jingqiny-99) [reply megatron/core/fusions/fused_mhc_kernels.py]: Fixed

File: megatron/core/transformer/hyper_connection.py [hunk 3]
```diff
@@ -15,95 +15,104 @@
...
+            M.backward(grad_output)
+        return logits.grad, None, None
 
-        Iteratively applies row and column normalization to project M
-        onto the Birkhoff polytope (doubly stochastic matrices).
 
-        Args:
-            M: [s, b, n, n] - positive matrix to normalize
-            num_iterations: Number of Sinkhorn iterations
+def ref_sinkhorn(input_logits: Tensor, num_iterations: int, eps: float = 1e-6) -> Tensor:
```
Comment (hxbai) [inline megatron/core/transformer/hyper_connection.py]: I think it is better to use a name with a prefix `native_`, `unfused_`, or something else. `ref_` looks like a unittest helper. The ones below are the same.

Comment (jingqiny-99) [reply megatron/core/transformer/hyper_connection.py]: Ok

File: megatron/core/transformer/hyper_connection.py [hunk 4]
```diff
@@ -15,95 +15,104 @@
...
-            # T_c: Column normalization
-            M = M / M.sum(dim=-2, keepdim=True).clamp(min=SinkhornKnopp.eps)
-        return M
 
-    @staticmethod
-    def forward(ctx, H_res_logits: Tensor, num_iterations: int) -> Tensor:
-        """
-        Project to doubly stochastic matrix via iterative row/col normalization.
+class NativeHAggregate(nn.Module):
+    """Native n-stream weighted aggregation: out = sum_j(h_pre_j * x_j)."""
```
Comment (Victarry) [inline megatron/core/transformer/hyper_connection.py:63]: Could we just write a plain python function for aggregate? Not sure about why need to define a new nn.Module.

File: megatron/core/transformer/hyper_connection.py [hunk 5]
```diff
@@ -15,95 +15,104 @@
     from megatron.core.tensor_parallel.random import CheckpointManager
 
 
-class SinkhornKnopp(torch.autograd.Function):
-    """
-    Differentiable Sinkhorn-Knopp algorithm for doubly stochastic projection.
+# ============================================================================
+# Reference implementations – forward only (autograd handles backward)
+# ============================================================================
```
Comment (Victarry) [inline megatron/core/transformer/hyper_connection.py]: This comment could just be removed.

File: megatron/core/transformer/hyper_connection.py [hunk 6]
```diff
@@ -15,95 +15,104 @@
...
+    M = torch.exp(input_logits - row_max)
+    for _ in range(num_iterations):
+        M = M / M.sum(dim=-1, keepdim=True).clamp(min=eps)
+        M = M / M.sum(dim=-2, keepdim=True).clamp(min=eps)
+    return M
 
-    Reference: Eq. (9) in mHC paper - M^{(t)} = T_c(T_r(M^{(t-1)}))
+
+class SinkhornKnopp(torch.autograd.Function):
+    """Reference Sinkhorn-Knopp projection to doubly stochastic matrix.
```
Comment (Victarry) [inline megatron/core/transformer/hyper_connection.py]: "Reference" to be deleted.

File: megatron/core/transformer/hyper_connection.py (hunk 4 above)
Comment (jingqiny-99) [reply megatron/core/transformer/hyper_connection.py:63]: Sure, I'll use function instead of torch module

**Comment by hxbai** [comment]
@jingqiny-99 Could you please attach the kernel performance numbers (and the comparison with native implementations) on the representative input sizes? Thanks.
